### PR TITLE
Bugfix: Propagate RHS type of resolved assignments

### PR DESCRIPTION
**Pull request by xermicus**
When resolving an assignment, the resolved variable should inherit the RHS type to avoid various issues in later compilation stages (for example with assignment chains).

### CODE REVIEW

File: src/codegen/expression.rs [hunk 1]
```diff
@@ -2667,7 +2667,7 @@ pub fn assign_single(
 
             Expression::Variable {
                 loc: left.loc(),
-                ty: ty.clone(),
+                ty: ty.deref_storage().clone(), // LHS value propagation is always in memory
                 var_no: pos,
```
Comment (seanyoung) [inline src/codegen/expression.rs:2671]: Although this works there is a mismatch between how variable `pos` is used, and how it is created on line 2584.

`--emit cfg --no-constant-folding --no-cse`
```
# function C::C::function::f__uint64 public:true selector:70e8dbcb nonpayable:true
# params: uint64 x
# returns: uint64
block0: # entry
	ty:uint64 %x = (arg #0)
	ty:uint64 storage %temp.11 = %x
	store storage slot(uint256 1) ty:uint64 = %temp.11
	ty:uint64 storage %temp.12 = %temp.11
	store storage slot(uint256 0) ty:uint64 = %temp.12
	return %temp.12
```
See how `%temp.11` and `%temp.12` are used as `uint64` but the variable type is `uint64 storage`.

Comment (xermicus) [reply src/codegen/expression.rs:2671]: Hmm yeah, I already tried to switch to the RHS type, but that lead to a bunch of codegen test failures. Maybe those are wrong too

Comment (seanyoung) [reply src/codegen/expression.rs:2671]: I noticed that too, I think the codegen tests will need fixing up

**Comment by xermicus** [comment]
I changed the fix to use the RHS type of the assignment. Which seems to fix a bunch of cases (see codegen tests).

The `FIXME` added to the constant folding codegen test should go away after #1450. Unless there is easy fix for this, I think it deserves a dedicated fix (in the sense of a unused variable pass after codegen / until it converges). Otherwise the problem might only be partially fixed.

File: src/sema/types.rs [hunk 2]
```diff
@@ -2075,6 +2075,14 @@ impl Type {
         }
     }
 
+    /// If the type is StorageRef, get the underlying type
+    pub fn deref_storage(&self) -> &Self {
```
[the remark this answers is not in the document]
Comment (xermicus) [reply src/sema/types.rs]: Good catch, I've removed it, if it is needed in the future can just add it again

**Review by seanyoung — APPROVED** [review]

**Review by LucasSte — APPROVED** [review]

**Pull request merged**